# Exploring eBay Car Sales Data

Goals: To clean the data and analyze the included used car listings. 

Dataset:  used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset was originally scraped and uploaded to Kaggle by user orgesleka. The original dataset isn't available on Kaggle anymore, but you can find it here.

We've made a few modifications from the original dataset. We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment

The data dictionary provided with data is as follows:

`dateCrawled` - When this ad was first crawled. All field-values are taken from this date.

`name` - Name of the car.

`seller` - Whether the seller is private or a dealer.

`offerType` - The type of listing

`price` - The price on the ad to sell the car.

`abtest` - Whether the listing is included in an A/B test.

`vehicleType` - The vehicle Type.

`yearOfRegistration` - The year in which which year the car was first registered.

`gearbox` - The transmission type.

`powerPS` - The power of the car in PS.

`model` - The car model name.

`kilometer` - How many kilometers the car has driven.

`monthOfRegistration` - The month in which which year the car was first registered.

`fuelType` - What type of fuel the car uses.

`brand` - The brand of the car.

`notRepairedDamage` - If the car has a damage which is not yet repaired.

`dateCreated` - The date on which the eBay listing was created.

`nrOfPictures` - The number of pictures in the ad.

`postalCode` - The postal code for the location of the vehicle.

`lastSeenOnline` - When the crawler saw this ad last online.



In [1]:
# Import pandas and NumPy libraries
import pandas as pd
import numpy as np

In [2]:
# Read dataset `autos.csv`
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Dataset Observations

According to `autos.info()`, the dataset is `50000 rows X 20 cols`, most of the column datatype is Object (String) except 5 columns are int type. 
Not-null values are mostly 50k, on the contrary, some columns exist fewer not-null values might need to clean up later, such as `vehicleType`, `gearbox`, `model`, `fuelType`, `notRepairedDamage`

For `autos.head()`, we notice some columns needs for data cleaning. 
1. `dateCrawled`: cast to timeframe data object
2. `name` : extract more meaning info such as brand name
3. `price`, `odometer`: cast to int or float data type for analysis

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

## Cleaning Columns Names

For all column names, change from `camelcase` to Python's preferred `snakecase`

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
new_cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']

autos.columns = new_cols
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Exploration and Cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

We'd like to drop columns having mostly one value.
For example, `seller`, `offer_type`, `abtest`, `gearbox`, `unrepaired_damage`

In [9]:
autos = autos.drop(['seller','offer_type','abtest','gearbox','unrepaired_damage'], axis=1)
autos.describe(include='all')

date_crawled         name  price vehicle_type  \
count                 50000        50000  50000        44905   
unique                48213        38754   2357            8   
top     2016-03-11 22:38:16  Ford_Fiesta     $0    limousine   
freq                      3           78   1421        12859   
mean                    NaN          NaN    NaN          NaN   
std                     NaN          NaN    NaN          NaN   
min                     NaN          NaN    NaN          NaN   
25%                     NaN          NaN    NaN          NaN   
50%                     NaN          NaN    NaN          NaN   
75%                     NaN          NaN    NaN          NaN   
max                     NaN          NaN    NaN          NaN   

        registration_year      power_ps  model   odometer  registration_month  \
count        50000.000000  50000.000000  47242      50000        50000.000000   
unique                NaN           NaN    245         13                 NaN   
top                   NaN           NaN   golf  150,000km                 NaN   
freq                  NaN           NaN   4024      32424                 NaN   
mean          2005.073280    116.355920    NaN        NaN            5.723360   
std            105.712813    209.216627    NaN        NaN            3.711984   
min           1000.000000      0.000000    NaN        NaN            0.000000   
25%           1999.000000     70.000000    NaN        NaN            3.000000   
50%           2003.000000    105.000000    NaN        NaN            6.000000   
75%           2008.000000    150.000000    NaN        NaN            9.000000   
max           9999.000000  17700.000000    NaN        NaN           12.000000   

       fuel_type       brand           ad_created  num_pictures   postal_code  \
count      45518       50000                50000       50000.0  50000.000000   
unique         7          40                   76           NaN           NaN   
top       benzin  volkswagen  2016-04-03 00:00:00           NaN           NaN   
freq       30107       10687                 1946           NaN           NaN   
mean         NaN         NaN                  NaN           0.0  50813.627300   
std          NaN         NaN                  NaN           0.0  25779.747957   
min          NaN         NaN                  NaN           0.0   1067.000000   
25%          NaN         NaN                  NaN           0.0  30451.000000   
50%          NaN         NaN                  NaN           0.0  49577.000000   
75%          NaN         NaN                  NaN           0.0  71540.000000   
max          NaN         NaN                  NaN           0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [10]:
autos["vehicle_type"].value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

In [11]:
autos["vehicle_type"].head()

0           bus
1     limousine
2     limousine
3    kleinwagen
4         kombi
Name: vehicle_type, dtype: object

In [12]:
autos["registration_year"].value_counts()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64

In [13]:
autos["registration_year"].head()

0    2004
1    1997
2    2009
3    2007
4    2003
Name: registration_year, dtype: int64

In [14]:
autos["price"].describe()

count     50000
unique     2357
top          $0
freq       1421
Name: price, dtype: object

In [15]:
autos["price"].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [16]:
autos["odometer"].describe()

count         50000
unique           13
top       150,000km
freq          32424
Name: odometer, dtype: object

In [17]:
autos["odometer"].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

Columns needs more clean up:
- registration_year: max, min out of valid range 1990-2022
- odometer: cast to numeric data type 
- price: cast to numberic data type


In [18]:
# remove non-numeric chars
unwanted_char = [",","$","km"]
for c in unwanted_char:
    autos["price"] = autos["price"].str.replace(c, "")
    autos["odometer"] = autos["odometer"].str.replace(c, "")
        
# conver col to numeric dtype
autos["price"] = autos["price"].astype(int)
autos["odometer"] = autos["odometer"].astype(int)

# remanme col name
autos.rename({"price":"price_usd", "odometer":"odometer_km"}, axis="columns", inplace=True)

In [19]:
autos["price_usd"].head()
autos["price_usd"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [20]:
autos["odometer_km"].head()
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

## Exploring the Odometer and Price Columns

Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

In [21]:
autos["odometer_km"].unique().shape

(13,)

In [22]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [23]:
autos["odometer_km"].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

From above exploration, `odometer_km` has no outliers.
Let's use the same technieuqes to explore `price_usd`

In [24]:
autos["price_usd"].unique().shape

(2357,)

In [25]:
autos["price_usd"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [26]:
autos["price_usd"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_usd, dtype: int64

In [27]:
autos["price_usd"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_usd, dtype: int64

For `price_usd`, we observed its mean=10000, medium=3000. Given eBay is an auction website, so its typical to have price as 0 for initial bid price. For our analysis purpose, it's safe to remove price data not in the range 1 to 100k

In [28]:
price_filter = autos["price_usd"].between(1, 100e3)
autos["price_usd"] = autos.loc[price_filter]["price_usd"]

In [29]:
autos["price_usd"].describe()

count    48526.000000
mean      5755.574146
std       7514.501772
min          1.000000
25%       1200.000000
50%       3000.000000
75%       7450.000000
max      99900.000000
Name: price_usd, dtype: float64

In [30]:
autos["price_usd"].value_counts().sort_index(ascending=False).head(20)

99900.0    2
99000.0    2
98500.0    1
94999.0    1
93911.0    1
93000.0    2
89900.0    1
89000.0    1
88900.0    1
86500.0    1
85000.0    1
84997.0    1
84000.0    1
83000.0    1
82987.0    1
80000.0    3
79999.0    1
79980.0    1
79933.0    1
79500.0    1
Name: price_usd, dtype: int64

## Exploring the Date Columns

There are 5 columns that consists date values
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Currentently, date_crawled, ad_created, last_seen are in the String format by pandas
Let's convert them to the timeframe dtype

In [31]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We found that first 10 chars represent the date (e.g. 2016-03-26).

Let's generate a distirbution for date

In [32]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [33]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

In [34]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

As we saw above, `date_craled` and `last_seen` date range in between `2016-03-05` to `2016-04-07`, whereas `ad_created` has a few date outside the date range (e.g.2015-xx-xx).


In [35]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

we observed `registration_year` date range includes min = 1000, max=9999. 

- The minimum value is 1000, before cars were invented
- The maximum value is 9999, many years into the future
Need to clean them up

## Dealing with Incorrect Registration Year Data

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the `1900 - 2016` interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [36]:
year_filter = autos["registration_year"].between(1900, 2016)
autos["registration_year"] = autos[year_filter]["registration_year"]

In [37]:
autos["registration_year"].value_counts(normalize=True).sort_index(ascending=True)

1910.0    0.000187
1927.0    0.000021
1929.0    0.000021
1931.0    0.000021
1934.0    0.000042
            ...   
2012.0    0.027546
2013.0    0.016782
2014.0    0.013867
2015.0    0.008308
2016.0    0.027401
Name: registration_year, Length: 78, dtype: float64

In [38]:
autos["registration_year"].describe()

count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64

Most of the registration year are range between 1999 to 2016. Only a few lies in early 20th century

## Exploring Price by Brand

When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the brand column. Here's what the process looks like:

- Identify the unique values we want to aggregate by
- Create an empty dictionary to store our aggregate data
- Loop over the unique values, and for each: 

    - Subset the dataframe by the unique values 
    - Calculate the mean of whichever column we're interested in 
    - Assign the val/mean to the dict as k/v.

In [39]:
autos["brand"].value_counts()

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
citroen             701
smart               701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
honda               399
kia                 356
alfa_romeo          329
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                110
subaru              109
land_rover           99
saab                 80
daewoo               79
trabant              78
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

To observe the trend, let's focus on the `top 20 brand` by number

In [40]:
top_20 = autos["brand"].value_counts().index[0:20]
print(top_20)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini'],
      dtype='object')


In [41]:
top_brand = {}
for brand in top_20:
    brand_filter = autos["brand"] == brand
    avg_price = autos[brand_filter]["price_usd"].mean()
    top_brand[brand] = avg_price

print(top_brand)

{'volkswagen': 5332.4784249226, 'opel': 2944.6075421641085, 'bmw': 8133.472016695125, 'mercedes_benz': 8440.438588943858, 'audi': 9173.144948404128, 'ford': 3691.064773735581, 'renault': 2431.195698924731, 'peugeot': 3065.611888111888, 'fiat': 2793.8700475435817, 'seat': 4315.744565217391, 'skoda': 6353.544871794872, 'mazda': 4059.059539918809, 'nissan': 4669.3859649122805, 'citroen': 3761.8950437317785, 'smart': 3518.102305475504, 'toyota': 5148.0032733224225, 'sonstige_autos': 10416.008583690988, 'hyundai': 5371.792960662526, 'volvo': 4866.993166287016, 'mini': 10541.566985645934}


In [42]:
top_brand_sorted = sorted(top_brand.items(), key=lambda x: x[1], reverse=True)

print(top_brand_sorted)

[('mini', 10541.566985645934), ('sonstige_autos', 10416.008583690988), ('audi', 9173.144948404128), ('mercedes_benz', 8440.438588943858), ('bmw', 8133.472016695125), ('skoda', 6353.544871794872), ('hyundai', 5371.792960662526), ('volkswagen', 5332.4784249226), ('toyota', 5148.0032733224225), ('volvo', 4866.993166287016), ('nissan', 4669.3859649122805), ('seat', 4315.744565217391), ('mazda', 4059.059539918809), ('citroen', 3761.8950437317785), ('ford', 3691.064773735581), ('smart', 3518.102305475504), ('peugeot', 3065.611888111888), ('opel', 2944.6075421641085), ('fiat', 2793.8700475435817), ('renault', 2431.195698924731)]


We observed top 6 brand in average price is:
`sonstige_autos`, `mini`, `audi`, `mercedes_benz`, `bmw`, `skoda`

## Storing Aggregate Data in a DataFrame

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

- it's difficult to compare more than two aggregate series objects if we want to extend to more columns
- we can't compare more than a few rows from each series object
- we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. To do this, we'll need to apply two pandas methods: 
- pandas series constructor
- pandas dataframe constructor

Use the loop method from the last screen to calculate the mean mileage and mean price for each of the top brands, storing the results in a dictionary.

In [43]:
top_brand_price = {}
top_brand_mileage = {} 
for brand in top_20:
    brand_filter = autos["brand"] == brand
    avg_price = autos[brand_filter]["price_usd"].mean()
    avg_mileage = autos[brand_filter]["odometer_km"].mean()
    top_brand_price[brand] = avg_price
    top_brand_mileage[brand] = avg_mileage

print(top_brand_price)

{'volkswagen': 5332.4784249226, 'opel': 2944.6075421641085, 'bmw': 8133.472016695125, 'mercedes_benz': 8440.438588943858, 'audi': 9173.144948404128, 'ford': 3691.064773735581, 'renault': 2431.195698924731, 'peugeot': 3065.611888111888, 'fiat': 2793.8700475435817, 'seat': 4315.744565217391, 'skoda': 6353.544871794872, 'mazda': 4059.059539918809, 'nissan': 4669.3859649122805, 'citroen': 3761.8950437317785, 'smart': 3518.102305475504, 'toyota': 5148.0032733224225, 'sonstige_autos': 10416.008583690988, 'hyundai': 5371.792960662526, 'volvo': 4866.993166287016, 'mini': 10541.566985645934}


In [44]:
print(top_brand_mileage)

{'volkswagen': 128955.27276129878, 'opel': 129298.66324848929, 'bmw': 132521.64302818198, 'mercedes_benz': 130886.14279678918, 'audi': 129643.9411627364, 'ford': 124131.93446392642, 'renault': 128223.79367720465, 'peugeot': 127352.33516483517, 'fiat': 117037.4617737003, 'seat': 122061.63655685441, 'skoda': 110947.83715012722, 'mazda': 125132.10039630119, 'nissan': 118978.7798408488, 'citroen': 119764.62196861627, 'smart': 100756.06276747503, 'toyota': 115988.65478119935, 'sonstige_autos': 87188.6446886447, 'hyundai': 106782.7868852459, 'volvo': 138632.3851203501, 'mini': 89375.0}


In [45]:
bmp_series = pd.Series(top_brand_price)
print(bmp_series)

volkswagen         5332.478425
opel               2944.607542
bmw                8133.472017
mercedes_benz      8440.438589
audi               9173.144948
ford               3691.064774
renault            2431.195699
peugeot            3065.611888
fiat               2793.870048
seat               4315.744565
skoda              6353.544872
mazda              4059.059540
nissan             4669.385965
citroen            3761.895044
smart              3518.102305
toyota             5148.003273
sonstige_autos    10416.008584
hyundai            5371.792961
volvo              4866.993166
mini              10541.566986
dtype: float64


In [46]:
bmm_series = pd.Series(top_brand_mileage)
print(bmm_series)

volkswagen        128955.272761
opel              129298.663248
bmw               132521.643028
mercedes_benz     130886.142797
audi              129643.941163
ford              124131.934464
renault           128223.793677
peugeot           127352.335165
fiat              117037.461774
seat              122061.636557
skoda             110947.837150
mazda             125132.100396
nissan            118978.779841
citroen           119764.621969
smart             100756.062767
toyota            115988.654781
sonstige_autos     87188.644689
hyundai           106782.786885
volvo             138632.385120
mini               89375.000000
dtype: float64


In [47]:
df_price = pd.DataFrame(bmp_series, columns=["mean_price"])
df_price

mean_price
volkswagen       5332.478425
opel             2944.607542
bmw              8133.472017
mercedes_benz    8440.438589
audi             9173.144948
ford             3691.064774
renault          2431.195699
peugeot          3065.611888
fiat             2793.870048
seat             4315.744565
skoda            6353.544872
mazda            4059.059540
nissan           4669.385965
citroen          3761.895044
smart            3518.102305
toyota           5148.003273
sonstige_autos  10416.008584
hyundai          5371.792961
volvo            4866.993166
mini            10541.566986

In [48]:
df_mileage = pd.DataFrame(bmm_series, columns=["mean_mileage"])
df_mileage

mean_mileage
volkswagen      128955.272761
opel            129298.663248
bmw             132521.643028
mercedes_benz   130886.142797
audi            129643.941163
ford            124131.934464
renault         128223.793677
peugeot         127352.335165
fiat            117037.461774
seat            122061.636557
skoda           110947.837150
mazda           125132.100396
nissan          118978.779841
citroen         119764.621969
smart           100756.062767
toyota          115988.654781
sonstige_autos   87188.644689
hyundai         106782.786885
volvo           138632.385120
mini             89375.000000

In [49]:
# Combine two dataframe together
df = df_price
df["mean_mileage"] = df_mileage
df

mean_price   mean_mileage
volkswagen       5332.478425  128955.272761
opel             2944.607542  129298.663248
bmw              8133.472017  132521.643028
mercedes_benz    8440.438589  130886.142797
audi             9173.144948  129643.941163
ford             3691.064774  124131.934464
renault          2431.195699  128223.793677
peugeot          3065.611888  127352.335165
fiat             2793.870048  117037.461774
seat             4315.744565  122061.636557
skoda            6353.544872  110947.837150
mazda            4059.059540  125132.100396
nissan           4669.385965  118978.779841
citroen          3761.895044  119764.621969
smart            3518.102305  100756.062767
toyota           5148.003273  115988.654781
sonstige_autos  10416.008584   87188.644689
hyundai          5371.792961  106782.786885
volvo            4866.993166  138632.385120
mini            10541.566986   89375.000000

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage. 